Requirements.md

```diff
@@ -0,0 +1,101 @@
+   
+   Within these languages, there are languages which are (largely) statically typed -- such as Java and C# -- and languages that are dynamically typed -- such as Python. We prioriterize statically typed languages over dynamically typed language; because those languages are often more performance sensitive, and because support for statically typed languages can often be used for dynamically typed languages.
+   
+1. Security is more important than architectural integrity.
+
+1. Architectural integrity is more important than performance.
+   'Architectural Integrity' is intended to denote simplicity/coherence of design. Generally speaking, architectural integrity can be measured in terms of the number of concepts and the complexity of their relationships.
+   
+   The more complex an architecture is, the harder it will be to extend or to modify; this, in turn, inhibits our ability to extend WASM-GC with new features.
+   
+   We prioriterize simple, composeable architectures over solution-oriented architectures. This will help to ensure that WASM-GC can evolove and be part of future WASM evolution.
+   
+   Our reasoning here is that performance is often the result of progressive tuning of an implementation. It can typically be improved over time. On the other hand, architectural deficiencies are very difficult to correct.
+      
+1. Performance is more important than convenience.
+
+   Since WASM-GC is not intended as a user-facing technology, convenience here refers to the ease with which toolchain authors can target WASM-GC. 
+   
+   Convenience for toolchain authors is still important however, as that affects adoption of WASM-GC. 
+
+   
+## Critical Success Factors
+
+A critical success factor is an aspect or property of possible solutions that may or may not be directly focused on the primary objective, but none-the-less is estimated to be crucially important to the success of the effort.
+
+1. Permit ‘cycle detection’ between host structures and language structures; so that such cycles can be collected if there are no other references to them.
+
+   When a WASM-GC module accesses host capabilities, or when a host application access structures from a WASM-GC library, cross references between them are likely to be established. For example, a DOM node may have an event listener that is actually implemented in WASM. The event listener, in turn, may have a reference to the DOM node it is listening to. This represents a cycle between the host and WASM-GC. If there are no other references to this cycles (if, for example, the DOM node is removed) then the entire cycle must be able to be collected.
+
+1. Performance implications of WASM-GC code should be straightforward.
```

Comment:
I think it is uncontroversial that as much optimization as possible should happen in toolchains rather than in engines. That being said, in practice different engines will have different performance profiles based on their complexity budget. For example, interpreters will always be slower and simpler than streaming compilers, which will always be slower and simpler than optimizing compilers.

The open question here is whether we want to allow for engines that have a higher complexity budget than the Web engines to be able to attain higher performance than web engines by e.g. allowing optimizing toolchains to provide low-level memory layout or tagging scheme hints that would be ignored by Web engines.

Reply:
I think it may be worth drilling into the term "performance" a bit, as it clearly has several aspects, including at least:

- "baseline performance", i.e. the performance that an engine with a low complexity budget might achieve, or the startup latency focused first tier of an optimizing multi-tier engine, or the fallback case of a selectively optimizing engine.
- "peak performance", i.e. the maximum performance that a complex, optimizing engine might achieve for long-running code, using strategies such as textbook compiler optimizations, dynamic feedback collection and JIT specialization, etc.
- "predictable performance", i.e. lack of performance variance across different engines, or for "minor" changes to the Wasm code, or other factors. A very strict version of this is often called "constant performance", which really _prevents_ engines from applying optimizations even if they could.

Different design decisions might impact only one of these aspects, or benefit one at the cost of another. Also, note that the lines can be a bit blurry, and/or there can be some confusion: someone who says "I want predictable performance" might actually mean "I want good baseline performance".

Which aspect does "performance implications should be straightforward" refer to?

In various discussions here the principle "Wasm should not require JIT-compilation to achieve good performance" has been brought up (the paragraph below about "special implementation techniques" may be a generalization of that notion). Personally, I'd assume that this specifically means "... good _baseline_ performance", and that it may be quite acceptable to have features where current state of the art is to achieve maximum performance for them by using adaptive JIT-compilation, as long as non-JIT baseline performance is also decent -- though I am aware that this assumption might not reflect wider consensus, so might be a good candidate for clarifying in this Requirements document.

Reply:
Personally, I've been assuming that we are designing to optimize peak performance for optimizing engines that will not do dynamic feedback collection. I agree that it would be helpful to check for consensus on what design constraints we are assuming for the top-tier of engines we are designing for.